### PR TITLE
Latency logging 2

### PR DESCRIPTION
**Pull request by lukaspetersson**

### CODE REVIEW

**Comment by lukaspetersson** [comment]
Use cloudlog.timestamp / LOGT with just a string to record a timestamp.
Use cloudlog.timestampExtra/LOGTE to log additional info, atm this is used in 3 places:
* Log translation between logMonoTime and frame id (sendcan does not have frameId field)
* Log the end of the pipeline
* Log submaster updated (start/end times + internal durations)

The frameId of controls is considered updated when it has received instructions from both the long and lat planner.

File: selfdrive/camerad/cameras/camera_qcom2.cc [hunk 1]
```diff
@@ -1032,8 +1032,10 @@ void process_road_camera(MultiCameraState *s, CameraState *c, int cnt) {
   if ((c == &s->road_cam && env_send_road) || (c == &s->wide_road_cam && env_send_wide_road)) {
     framed.setImage(get_frame_image(b));
   }
+  LOGT((std::string(c == &s->road_cam ? "RoadCamera" : "WideRoadCamera")+std::string(": Image set")).c_str());
```
Comment (adeebshihadeh) [inline selfdrive/camerad/cameras/camera_qcom2.cc:1035]: can you build these strings in a nicer way?

File: tools/latency_logging/print_timestamps.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+import json
+from collections import defaultdict
+
+r = Route("9f583b1d93915c31|2022-03-28--17-25-12") 
+lr = LogReader(r.log_paths()[0])
+
+services = ['camerad', 'modeld', 'plannerd', 'controlsd', 'boardd']
+
+msgq_to_service = {}
+msgq_to_service['roadCameraState'] = "camerad"
```
Comment (adeebshihadeh) [inline tools/latency_logging/print_timestamps.py:13]:
```suggestion
msgq_to_service = {
  'roadCameraState': 'camerad',
  'modelV2': 'modeld',
  ...
}
```

File: selfdrive/controls/controlsd.py [hunk 3]
```diff
@@ -624,10 +627,13 @@ def publish_logs(self, CS, start_time, CC, lac_log):
     if current_alert:
       hudControl.visualAlert = current_alert.visual_alert
 
+    frame_id = CC.frameId
     if not self.read_only and self.initialized:
       # send car controls over can
       self.last_actuators, can_sends = self.CI.apply(CC)
-      self.pm.send('sendcan', can_list_to_can_capnp(can_sends, msgtype='sendcan', valid=CS.canValid))
+      sendcan_bytes = can_list_to_can_capnp(can_sends, msgtype='sendcan', valid=CS.canValid)
+      cloudlog.timestampExtra("translation", {"from": log.Event.from_bytes(sendcan_bytes).logMonoTime, "to": frame_id})
```
Comment (adeebshihadeh) [inline selfdrive/controls/controlsd.py:635]: why is this done here? this LUT can be recovered from the rest of the logs and implementation details like this shouldn't leak over here

File: selfdrive/controls/lib/lateral_planner.py [hunk 4]
```diff
@@ -103,6 +103,7 @@ def publish(self, sm, pm):
     plan_send.valid = sm.all_alive_and_valid(service_list=['carState', 'controlsState', 'modelV2'])
 
     lateralPlan = plan_send.lateralPlan
+    lateralPlan.frameId = sm['modelV2'].frameId
```
Comment (adeebshihadeh) [inline selfdrive/controls/lib/lateral_planner.py:106]: let's use the logMonoTime like longitudinalPlan does

File: tools/latency_logging/print_timestamps.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+                    skip_frames.append(frame_id)
+            for duration in service_to_durations[service]:
+                if duration == 'timestampSof':
+                    service_blocks[int(frame_id)][service]["Start"].append(float(smInfo[duration]))
+                else:
+                    durations[int(frame_id)][service][duration].append(smInfo[duration]*1e3)
+            service_blocks[frame_id][service]["End"].append(float(pubTime))
+            next_service = services[services.index(service)+1]
+            service_blocks[int(frame_id)][next_service]["Start"].append( float(rcvTime))
+        elif "Pipeline end" in msg:
```
Comment (adeebshihadeh) [inline tools/latency_logging/print_timestamps.py:72]: why is this needed?

File: tools/latency_logging/print_timestamps.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+service_to_durations = defaultdict(list)
+service_to_durations['camerad'] = ['timestampSof', 'processingTime']
+service_to_durations['modeld'] = ['modelExecutionTime', 'gpuExecutionTime']
+service_to_durations['plannerd'] = ["solverExecutionTime"]
+
+# Build translation dict
+translationdict = {}
+for msg in lr:
+  if msg.which() == "logMessage" and "translation" in msg.logMessage:
+    msg = msg.logMessage.replace("'", '"').replace('"{', "{").replace('}"', "}").replace("\\", '')
```
Comment (adeebshihadeh) [inline tools/latency_logging/print_timestamps.py:29]: looks like this is just a string-ified python dict instead of properly serialized json

File: selfdrive/manager/manager.py [hunk 7]
```diff
@@ -169,6 +172,19 @@ def manager_thread() -> None:
     if shutdown:
       break
 
+def logTimestamps(sm, msgqs):
```
Comment (adeebshihadeh) [inline selfdrive/manager/manager.py:175]: this should be recoverable from the logs

**Review by adeebshihadeh — COMMENTED** [review]
cloudlog.timestamp is looking pretty nice now that we're just passing a string! Make sure that all CI is passing and the PR is out of draft before requesting a review.

File: common/logging_extra.py [hunk 8]
```diff
@@ -11,6 +11,13 @@
 from collections import OrderedDict
 from contextlib import contextmanager
 
+try:
+    from common.realtime import sec_since_boot
```
[the remark this answers is not in the document]
Comment (lukaspetersson) [reply common/logging_extra.py:15]: done

File: common/logging_extra.py [hunk 9]
```diff
@@ -11,6 +11,13 @@
...
+except ImportError:
+    import time
+    sec_since_boot = time.time
```
[the remark this answers is not in the document]
Comment (lukaspetersson) [reply common/logging_extra.py:18]: done

File: selfdrive/common/swaglog.cc [hunk 10]
```diff
@@ -8,9 +8,9 @@
 #include <cstring>
 #include <mutex>
 #include <string>
+#include <chrono>
```
[the remark this answers is not in the document]
Comment (lukaspetersson) [reply selfdrive/common/swaglog.cc:11]: done

File: selfdrive/common/swaglog.h [hunk 11]
```diff
@@ -13,10 +15,32 @@
...
+  uint64_t ns = nanos_since_boot();                               \
+  json11::Json tspt_e_j = json11::Json::object {                    \
+    {"timestampExtra", json11::Json::object{{"event", event_name},     \
+                   {"time", std::to_string(ns)},                  \
+                   {"info", info}}                  \
+    }};                                                           \
+  cloudlog(lvl, tspt_e_j.dump().c_str(), ## __VA_ARGS__);           \
+}
+
+#define cloudlog_t(lvl, event_name, ...)                           \
```
[the remark this answers is not in the document]
Comment (lukaspetersson) [reply selfdrive/common/swaglog.h:34]: done

File: selfdrive/modeld/modeld.cc [hunk 12]
```diff
@@ -89,7 +89,6 @@ void run_model(ModelState &model, VisionIpcClient &vipc_client_main, VisionIpcCl
       LOGE("vipc_client_main no frame");
       continue;
     }
-
```
[the remark this answers is not in the document]
Comment (lukaspetersson) [reply selfdrive/modeld/modeld.cc:92]: ok

File: selfdrive/controls/lib/longitudinal_planner.py [hunk 13]
```diff
@@ -116,6 +116,7 @@ def publish(self, sm, pm):
     plan_send.valid = sm.all_alive_and_valid(service_list=['carState', 'controlsState'])
 
     longitudinalPlan = plan_send.longitudinalPlan
+    longitudinalPlan.frameId = sm['modelV2'].frameId
```
[the remark this answers is not in the document]
Comment (lukaspetersson) [reply selfdrive/controls/lib/longitudinal_planner.py:119]: But then a translation is needed which is even more complex.

I could skip tracking the frame id of longitudinal and only track the path through lateral. But I think it is nice to do both.

File: selfdrive/common/swaglog.h [hunk 14]
```diff
@@ -1,7 +1,9 @@
 #pragma once
 
 #include "selfdrive/common/timing.h"
+#include "json11.hpp"
 
+#define CLOUDLOG_TIMESTAMP 10
```
[the remark this answers is not in the document]
Comment (lukaspetersson) [reply selfdrive/common/swaglog.h:6]: done

File: selfdrive/controls/controlsd.py [hunk 15]
```diff
@@ -486,6 +486,9 @@ def state_control(self, CS):
     long_plan = self.sm['longitudinalPlan']
 
     CC = car.CarControl.new_message()
+    # consider frameId updated when both are updated
+    frame_id = min(lat_plan.frameId, long_plan.frameId)
+    CC.frameId = frame_id
```
[the remark this answers is not in the document]
Comment (lukaspetersson) [reply selfdrive/controls/controlsd.py:491]: mistake from past, done

File: selfdrive/controls/controlsd.py [hunk 16]
```diff
@@ -721,9 +727,11 @@ def step(self):
 
     # Sample data from sockets and get a carState
     CS = self.data_sample()
+    cloudlog.timestamp("Data sampled")
     self.prof.checkpoint("Sample")
```
[the remark this answers is not in the document]
Comment (lukaspetersson) [reply selfdrive/controls/controlsd.py:731]: ok

File: selfdrive/controls/controlsd.py (hunk 3 above)
Comment (lukaspetersson) [reply selfdrive/controls/controlsd.py:633]: How do you bind the logMono to the frame Id in the logs? You would get a planner message with frame Id and later a sendcan with logMono. But how do we know that these correspond to the same?

If the answer is that we can assume that the message after a planner message is a controls message corresponding to the same frame, why do we track the frame id at all? Cant we just assume the repeated sequence camera-model-planner-controls-board

File: selfdrive/camerad/cameras/camera_qcom2.cc (hunk 1 above)
Comment (lukaspetersson) [reply selfdrive/camerad/cameras/camera_qcom2.cc:1035]: Tried to use va_args to allow for an unknown number of c-str arguments, but got stuck

File: selfdrive/manager/manager.py (hunk 7 above)
Comment (lukaspetersson) [reply selfdrive/manager/manager.py:175]: But this is were it is inserted in the logs (?). I tried to find this info in the logs without this function but could not find it.

File: tools/latency_logging/print_timestamps.py [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+msgq_to_service['wideRoadCameraState'] = "camerad"
+msgq_to_service['modelV2'] = "modeld"
+msgq_to_service['lateralPlan'] = "plannerd"
+msgq_to_service['longitudinalPlan'] = "plannerd"
+msgq_to_service['sendcan'] = "controlsd"
+
+service_to_durations = defaultdict(list)
+service_to_durations['camerad'] = ['timestampSof', 'processingTime']
+service_to_durations['modeld'] = ['modelExecutionTime', 'gpuExecutionTime']
+service_to_durations['plannerd'] = ["solverExecutionTime"]
```
[the remark this answers is not in the document]
Comment (lukaspetersson) [reply tools/latency_logging/print_timestamps.py:23]: Why? That would be O(n) instead of O(1) for lookup

File: tools/latency_logging/print_timestamps.py (hunk 6 above)
Comment (lukaspetersson) [reply tools/latency_logging/print_timestamps.py:33]: When printing msg:

( logMonoTime = 599335952933494,                                                                                            |i
  logMessage = "{\"created\": 1648513515.2507977, \"ctx\": {\"daemon\": \"camerad\", \"device\": \"tici\", \"dirty\": true, |o
ongle_id\": \"9f583b1d93915c31\", \"version\": \"0.8.14\"}, \"filename\": \"selfdrive/camerad/cameras/camera_qcom2.cc\", \"f|n
name\": \"process_road_camera\", \"levelnum\": 10, \"lineno\": 1035, \"msg\": \"{\\\"timestamp\\\": {\\\"event\\\": \\\"Road|s
era: Image set\\\", \\\"time\\\": \\\"599335951599136\\\"}}\"}",                                                            |
  valid = true )

could look into where the logs are set, but it is not only the logging which I have created

Comment (lukaspetersson) [reply tools/latency_logging/print_timestamps.py:33]: The cpp json11 lib causes this. Confused why it does that, using it exactly as the docs/

\"{\\\"timestampExtra\\\"|        'longitudinalPlan':'plannerd',
: {\\\"event\\\": \\\"Pipeline end\\\", \\\"i                                     |        'sendcan':'controlsd'
nfo\\\": {\\\"logMonoTime\\\": \\\"599335735751377\\\"}, \\\"time\\\": \\\"5993357|        }
36414806\\\"}}\"}"

File: selfdrive/controls/lib/lateral_planner.py (hunk 4 above)
Comment (lukaspetersson) [reply selfdrive/controls/lib/lateral_planner.py:106]: Use translations instead? I think this is worse

File: tools/latency_logging/print_timestamps.py (hunk 5 above)
Comment (lukaspetersson) [reply tools/latency_logging/print_timestamps.py]: Because sendcan is sent with usb. this is not logged.

File: selfdrive/manager/manager.py [hunk 18]
```diff
@@ -169,6 +172,19 @@ def manager_thread() -> None:
...
+  for msgq in msgqs:
+    if sm.updated[msgq]:
+      sm_info = {}
+      sm_info['msg_name'] = msgq
+      sm_info['logMonoTime'] = sm.logMonoTime[msgq]
+      sm_info['rcvTime'] = sm.rcv_time[msgq]
```
[the remark this answers is not in the document]
Comment (lukaspetersson) [reply selfdrive/manager/manager.py:182]: True

File: tools/latency_logging/print_timestamps.py [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+      for event, times in dict(events).items():
+        print("    "+'%-50s%-50s' %(event, str(times)))  
+
+def graph_timestamps(service_intervals, relative_self):
+  t0 = min([min([min(times) for times in events.values()]) for events in service_intervals[min(service_intervals.keys())].values()])
+  gnt = plt.subplots()[1]
+  gnt.set_xlim(0, 150 if relative_self else 750)
+  gnt.set_ylim(0, len(service_intervals))
+  y = 0
+  for frame_id, services in service_intervals.items():
```
Comment (adeebshihadeh) [inline tools/latency_logging/print_timestamps.py:166]: use enumerate instead
```suggestion
  for y, frame_id, services in enumerate(service_intervals.items()):
```

File: selfdrive/controls/plannerd.py [hunk 20]
```diff
@@ -37,8 +37,10 @@ def plannerd_thread(sm=None, pm=None):
 
     if sm.updated['modelV2']:
       lateral_planner.update(sm)
+      cloudlog.timestamp("Lateral planner updated")
```
Comment (adeebshihadeh) [inline selfdrive/controls/plannerd.py:40]: move these to before the update. the logMonoTime is basically determined right after this

File: selfdrive/modeld/models/driving.cc [hunk 21]
```diff
@@ -74,12 +75,14 @@ ModelOutput* model_eval_frame(ModelState* s, VisionBuf* buf, VisionBuf* wbuf,
   // if getInputBuf is not NULL, net_input_buf will be
   auto net_input_buf = s->frame->prepare(buf->buf_cl, buf->width, buf->height, transform, static_cast<cl_mem*>(s->m->getInputBuf()));
   s->m->addImage(net_input_buf, s->frame->buf_size);
+  LOGT("Image added");
```
Comment (adeebshihadeh) [inline selfdrive/modeld/models/driving.cc:78]: log addExtra as well

File: selfdrive/modeld/models/driving.cc [hunk 22]
```diff
@@ -74,12 +75,14 @@ ModelOutput* model_eval_frame(ModelState* s, VisionBuf* buf, VisionBuf* wbuf,
...
 
   if (wbuf != nullptr) {
     auto net_extra_buf = s->wide_frame->prepare(wbuf->buf_cl, wbuf->width, wbuf->height, transform_wide, static_cast<cl_mem*>(s->m->getExtraBuf()));
     s->m->addExtra(net_extra_buf, s->wide_frame->buf_size);
   }
   s->m->execute();
+  LOGT("Execution finnished");
```
Comment (adeebshihadeh) [inline selfdrive/modeld/models/driving.cc:85]: typo

File: tools/latency_logging/print_timestamps.py [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+      jmsg = json.loads(msg.logMessage)
+      if "timestamp" in jmsg['msg']:
+        cloudlog_logs.append(jmsg)
+  return (sm_logs, cloudlog_logs)
+
+def get_translation_LUT(sm_logs):
+  translationdict = {}
+  # sendcan is the only message where we cannot identify the frame id with ceirtainty, but assuing it has the same as the next constrolsstate is good enough (sm_logs are sorted by time)
+  latest_sendcan_monotime = 0
+  for msg in sm_logs:
```
Comment (adeebshihadeh) [inline tools/latency_logging/print_timestamps.py:49]: this can directly take the LogReader

File: tools/latency_logging/print_timestamps.py [hunk 24]
```diff
@@ -0,0 +1,254 @@
...
+      if msg.which() == 'modelV2':
+        translationdict[mono_time]=msg.modelV2.frameId
+      elif msg.which() == 'lateralPlan':
+        translationdict[mono_time]=translationdict[msg.lateralPlan.modelMonoTime]
+      elif msg.which() == 'sendcan':
+        latest_sendcan_monotime = msg.logMonoTime
+      elif msg.which() == 'controlsState':
+        if latest_sendcan_monotime:
+          translationdict[latest_sendcan_monotime]=translationdict[msg.controlsState.lateralPlanMonoTime]
+    except KeyError:
```
Comment (adeebshihadeh) [inline tools/latency_logging/print_timestamps.py:61]: why can this happen?

File: tools/latency_logging/print_timestamps.py [hunk 25]
```diff
@@ -0,0 +1,254 @@
...
+      mono_time = msg.logMonoTime
+      if msg.which() == 'modelV2':
+        translationdict[mono_time]=msg.modelV2.frameId
+      elif msg.which() == 'lateralPlan':
+        translationdict[mono_time]=translationdict[msg.lateralPlan.modelMonoTime]
+      elif msg.which() == 'sendcan':
+        latest_sendcan_monotime = msg.logMonoTime
+      elif msg.which() == 'controlsState':
+        if latest_sendcan_monotime:
+          translationdict[latest_sendcan_monotime]=translationdict[msg.controlsState.lateralPlanMonoTime]
```
Comment (adeebshihadeh) [inline tools/latency_logging/print_timestamps.py:60]: add whitespace around the equals

File: tools/latency_logging/print_timestamps.py [hunk 26]
```diff
@@ -0,0 +1,254 @@
...
+      sm_logs.append(msg)
+    elif msg.which() == "logMessage" and 'timestamp' in msg.logMessage:
+      jmsg = json.loads(msg.logMessage)
+      if "timestamp" in jmsg['msg']:
+        cloudlog_logs.append(jmsg)
+  return (sm_logs, cloudlog_logs)
+
+def get_translation_LUT(sm_logs):
+  translationdict = {}
+  # sendcan is the only message where we cannot identify the frame id with ceirtainty, but assuing it has the same as the next constrolsstate is good enough (sm_logs are sorted by time)
```
Comment (adeebshihadeh) [inline tools/latency_logging/print_timestamps.py:47]: typos

File: selfdrive/controls/plannerd.py (hunk 20 above)
Comment (lukaspetersson) [reply selfdrive/controls/plannerd.py:40]: True, but the previous log mono is determined right before if we move.

I think deleting is best, together with solverExecutionTime we have enough info

File: tools/latency_logging/print_timestamps.py [hunk 27]
```diff
@@ -0,0 +1,254 @@
...
+
+SERVICE_TO_DURATIONS = {
+        'camerad':['processingTime'],
+        'modeld':['modelExecutionTime', 'gpuExecutionTime'],
+        'plannerd':["solverExecutionTime"],
+        'controlsd':[],
+        'boardd':[]
+        }
+
+def get_relevant_logs(logreader):
```
[the remark this answers is not in the document]
Comment (lukaspetersson) [reply tools/latency_logging/print_timestamps.py:32]: The assumption is that the relevant logs is a small subset of all logs. Since I need to loop though them 3 times, this improves speed.

I could have used this loop as the first of the 3 loops, but to improve readability I assumed that en extra loop of the small subset was negligible.

File: tools/latency_logging/print_timestamps.py (hunk 24 above)
Comment (lukaspetersson) [reply tools/latency_logging/print_timestamps.py:61]: not happening anymore

File: selfdrive/common/swaglog.cc [hunk 28]
```diff
@@ -86,8 +86,12 @@ void cloudlog_e(int levelnum, const char* filename, int lineno, const char* func
 
   if (!s.initialized) s.initialize();
 
+  // if msg is json, parse it
+  std::string err;
+  auto json_msg = json11::Json::parse(msg_buf, err);
```
[the remark this answers is not in the document]
Comment (lukaspetersson) [reply selfdrive/common/swaglog.cc:91]: cloudlog_t sends a string from .dump(). this string must be parsed before added to another json string. Otherwise extra backslashes are added.

This was the best solution to the .replace("/", "") situation I found

File: tools/latencylogger/README.md [hunk 29]
```diff
@@ -0,0 +1,87 @@
+# LatencyLogger
+
+LatencyLogger is a tool to track timestamps for each frame in the pipeline. Timestamps are printed in a table as well as plotted in a graph. All times are in milliseconds.
```
Comment (adeebshihadeh) [inline tools/latencylogger/README.md:3]:
```suggestion
LatencyLogger is a tool to track the time from first pixel to actuation. Timestamps are printed in a table as well as plotted in a graph. Start openpilot with `LOG_TIMESTAMPS=1` set to enable the necessary logging.
```

File: tools/latencylogger/README.md [hunk 30]
```diff
@@ -0,0 +1,87 @@
...
+
+optional arguments:
+  -h, --help            show this help message and exit
+  --relative            Make timestamps relative to the start of each frame (default: False)
+  --demo                Use the demo route instead of providing one (default: False)
+  --plot                If a plot should be generated (default: False)
+```
+
+## Examples
+Plotting with relative_self starts each process at time=0 and gives a nice overview.
```
Comment (adeebshihadeh) [inline tools/latencylogger/README.md:25]: relative_self -> relative
